packages/airnode-protocol-v1/contracts/monetization/RequesterAuthorizerWhitelisterWithTokenDeposit.sol

```diff
@@ -158,6 +189,73 @@ contract RequesterAuthorizerWhitelisterWithTokenDeposit is
                     false
                 );
         }
+    }
+
+    /// @notice Withdraws tokens previously deposited for the requester to be
+    /// whitelisted for the Airnode–endpoint pair on the chain
+    /// @param airnode Airnode address
+    /// @param chainId Chain ID
+    /// @param endpointId Endpoint ID
+    /// @param requester Requester address
+    function withdrawTokens(
+        address airnode,
+        uint256 chainId,
+        bytes32 endpointId,
+        address requester
+    ) external override onlyNonBlockedRequester(airnode, requester) {
+        TokenDeposits
+            storage tokenDeposits = airnodeToChainIdToEndpointIdToRequesterToTokenDeposits[
+                airnode
+            ][chainId][endpointId][requester];
+        uint256 tokenWithdrawAmount = tokenDeposits.depositorToAmount[
+            msg.sender
+        ];
+        require(tokenWithdrawAmount != 0, "Sender has not deposited tokens");
+        tokenDeposits.depositorToAmount[msg.sender] = 0;
+        uint256 earliestWithdrawalTime = tokenDeposits
+            .depositorToEarliestWithdrawalTime[msg.sender];
+        if (withdrawalLeadTime != 0) {
+            require(
+                earliestWithdrawalTime != 0,
+                "Withdrawal intent not signaled"
+            );
+        }
+        if (earliestWithdrawalTime != 0) {
+            require(
+                earliestWithdrawalTime <= block.timestamp,
+                "Not withdrawal time yet"
+            );
+            uint256 tokenDepositsCount = tokenDeposits.count;
```

Comment:
I don't think we should be setting `tokenDeposits.count` in `signalWithdrawalIntent()` instead it should be done here
 ```
uint256 tokenDepositsCount = --tokenDeposits.count;
```

Reply:
This doesn't work because then the users that expect to be blocked will signal withdrawal intent immediately after depositing, and would be able to frontrun the blocking transaction.
The DAO staking pool takes away your voting power and stops you from receiving staking rewards as soon as you schedule an unstake (and not at the time that you actually unstake) for the same reason.

Reply:
I don't really see how that's a problem. Even if the user signals withdrawIntent() he won't be able to withdraw because of the onlyNonBlockedRequester modifier. We would have to withdraw the blocked tokens to revoke the whitelisting but I think that's fine.

Reply:
Also there is another problem with this approach, say for example a user is blocked and he signals `withdrawIntent()` so `--tokenDeposits.count` happens once. If we call `withdrawFundsDepositedForBlockedRequester()` it will trigger another `--tokenDeposits.count` for the same user.

Reply:
> I don't really see how that's a problem. Even if the user signals withdrawIntent() he won't be able to withdraw because of the onlyNonBlockedRequester modifier in withdrawTokens().

The only reason for this PR is that the call to withdrawTokens() can frontrun the blocking call. Read the audit report again.

> Also there is another problem with this approach...

Good point, I'll fix that

Reply:
Fixed in https://github.com/api3dao/airnode/pull/926/commits/91938b143cf14a0d8dcbd2b1d58101f26a51ce80

Reply:
reread your initial comment and I think I understand what you mean. I agree that won't work.

Reply:
> Fixed in [91938b1](https://github.com/api3dao/airnode/commit/91938b143cf14a0d8dcbd2b1d58101f26a51ce80)

we should probably update the tests to cover both the cases.

Reply:
Yeah, but I think we can do that later on. This contract will probably not be used in this form anyway and I have a lot of time-critical work.